Python/bltinmodule.c

```diff
@@ -2687,11 +2687,9 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 continue;
             }
             if (PyLong_Check(item)) {
-                long value;
-                int overflow;
-                value = PyLong_AsLongAndOverflow(item, &overflow);
-                if (!overflow) {
-                    re_sum.hi += (double)value;
+                double value = PyLong_AsDouble(item);
+                if (value != -1.0 || !PyErr_Occurred()) {
+                    re_sum = cs_add(re_sum, value);
                     Py_DECREF(item);
                     continue;
                 }
```

Comment:
We should be sure the exception gets handled:

```suggestion
                double value = PyLong_AsDouble(item);
                if (value != -1.0 || !PyErr_Occurred()) {
                    re_sum = cs_add(re_sum, value);
                    Py_DECREF(item);
                    continue;
                }
                if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
                    PyErr_Clear();
                }
                else {
                    return NULL;
                }
```

Reply:
Also, it would be good to add a test for this case: `item` overflows/underflows a `double`.

Reply:
Thanks!  I think we should add a quick exit (return NULL) as an alternative.

But I doubt we should use PyErr_ExceptionMatches.  Rather, we should exit unconditionally: [OverflowError](https://docs.python.org/3/library/exceptions.html#OverflowError) is *only* possible exception at this point.  And when we fallback to PyNumber_Add() - this exception will be raised again:
```pycon
Python 3.12.0+ (heads/3.12:2162512d71, Jan 13 2024, 11:25:20) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> sum([1.0, 10**1000])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: int too large to convert to float
```